magenta/music/musicxml_reader.py

```diff
@@ -0,0 +1,116 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""MusicXML import.
+
+Input wrappers for converting MusicXML into tensorflow.magenta.NoteSequence.
+"""
+
+# internal imports
+from magenta.music.musicxml_parser import MusicXMLDocument
+from magenta.protobuf import music_pb2
+# pylint: enable=g-import-not-at-top
+
+class MusicXMLConversionError(Exception):
+  """MusicXML conversion error handler"""
+  pass
+
+def musicxml_to_sequence_proto(musicxml_document):
+  """Convert MusicXML file contents to a tensorflow.magenta.NoteSequence proto.
+
+  Converts a MusicXML file encoded as a string into a
+  tensorflow.magenta.NoteSequence proto.
+
+  Args:
+    musicxml_document: A parsed MusicXML file.
+    This file has been parsed by class MusicXMLDocument
+
+  Returns:
+    A tensorflow.magenta.NoteSequence proto.
+
+  Raises:
+    MusicXMLConversionError: An error occurred when parsing the MusicXML file.
+  """
+
+  sequence = music_pb2.NoteSequence()
+
+  # Populate header.
+  sequence.ticks_per_quarter = musicxml_document.midi_resolution
+
+  # Populate time signatures.
+  musicxml_time_signatures = musicxml_document.gettimesignatures()
+  for musicxml_time_signature in musicxml_time_signatures:
+    time_signature = sequence.time_signatures.add()
+    time_signature.time = musicxml_time_signature.time_position
+    time_signature.numerator = musicxml_time_signature.numerator
+    time_signature.denominator = musicxml_time_signature.denominator
+
+  # Populate key signatures.
+  musicxml_key_signatures = musicxml_document.getkeysignatures()
+  for musicxml_key in musicxml_key_signatures:
+    key_signature = sequence.key_signatures.add()
+    key_signature.time = musicxml_key.time_position
+    # The Key enum in music.proto does NOT follow MIDI / MusicXML specs
+    # Convert from MIDI / MusicXML key to music.proto key
+    music_proto_keys = [11, 6, 1, 8, 3, 10, 5, 0, 7, 2, 9, 4, 11, 6, 1]
+    key_signature.key = music_proto_keys[musicxml_key.key + 7]
+    if musicxml_key.mode == "major":
+      key_signature.mode = key_signature.MAJOR
+    elif musicxml_key.mode == "minor":
+      key_signature.mode = key_signature.MINOR
+
+  # Populate tempo changes.
+  musicxml_tempos = musicxml_document.gettempos()
+  for musicxml_tempo in musicxml_tempos:
+    tempo = sequence.tempos.add()
+    tempo.time = musicxml_tempo.time_position
+    tempo.qpm = musicxml_tempo.qpm
+
+  # Populate notes from each MusicXML part across all voices
+  # Unlike MIDI import, notes are not sorted
```

Comment:
I didn't understand the rationale for sorting the MIDI notes, so I didn't sort the notes from the MusicXML. What purpose does sorting serve?